cmd/pebble/test.go

```diff
@@ -324,27 +326,33 @@ func runTest(dir string, t test) {
 	workersDone := make(chan struct{})
 	signal.Notify(done, os.Interrupt)
 
-	go func() {
+	eg, ctx := errgroup.WithContext(context.Background())
+	defer ctx.Done()
```

Comment:
I think you want to do a `context.WithCancel(context.Background())` and call the returned cancel function in a `defer` here, and create the `errgroup` out of that context. `ctx.Done` just returns a Done channel, it doesn't do anything on its own.

Also I don't think we ever call `eg.Wait()`, maybe that can happen in a defer before the cancel function so that it runs after the context is cancelled.

Reply:
oof. this initial implementation was a bit sloppy. Apologies!

So I ripped out the errgroup and I now ensure the `maxSize` and `duration` goroutines finish. I'm not sure how to guarantee that the first `workersDone` goroutine finishes before `runTest` terminates without adding a bunch complicated scaffolding: 

I suppose I could add in an `errgroup` with the other go routines and have the maxSize and duration goroutines send a non nil context cancellation error when either expire?

Reply:
This mostly looks good to me now! and no it shouldn't be complex scaffolding, all you need to do is create a `var wg sync.WaitGroup`, do a `defer wg.Wait()` on it before the `defer cancel()` (so it runs AFTER the cancel, as defers run in reverse order), and do the appropriate `wg.Add()`s and a `defer wg.Done()` in each of the child goroutines.

Reply:
hm, I think your proposal may cause the process to hang forever, but maybe i'm missing something. The current `wg.Wait()` call in the first child goroutine only returns if the user specifies the `--num-ops` flag, at least for `bench ycsb`. I.e. the goroutine below:
```
go func() {
		wg.Wait()
		close(workersDone)
	}()
```

In other words, your proposed `defer parentWaitGroup.Done()` in this goroutine would never get called, which means runTest could never return. Currently, I'm pretty sure the ycsb workers only shut down when the process gets killed. 

So, for you proposal to work, I think we'd need to build a cleaner shutdown of the worker goroutines after `t.init(db,&wg)` gets called.